include/revng/CliftTransforms/Expressions.td

```diff
@@ -0,0 +1,130 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#ifndef MLIR_CLIFT_EXPRESSION_REWRITES
+#define MLIR_CLIFT_EXPRESSION_REWRITES
+
+include "mlir/IR/PatternBase.td"
+
+include "revng/Clift/Clift.td"
+include "revng/CliftTransforms/ExpressionHelpers.td"
+
+//===------------------------- Boolean expressions -------------------------===//
+
+// x == 0 -> !x
+def CmpEqWithZeroPattern
+  : Pat<(Clift_CmpEqOp:$res $arg, (Clift_ImmediateOp $value)),
+        (Clift_LogicalNotOp $arg),
+        [(Constraint<CPred<"$0.getValue().isZero()">> $value)]>;
+
+// x != 0 -> x
+//           x ? 1 : 0
+// Note: The choice of replacement depends on whether the result is
+//       boolean-tested.
+def CmpNeWithZeroPattern
+  : Pat<(Clift_CmpNeOp:$res $arg, (Clift_ImmediateOp:$imm $value)),
+        (forceBooleanResult $res, $arg),
+        [(Constraint<CPred<"$0.getValue().isZero()">> $value)]>;
+
+//===---------------------------- Miscellaneous ---------------------------===//
+
+// (T)123 -> 123
+// Constraint: T is a primitive integer type
+def CastImmediatePattern
+  : Pat<(Clift_CastOp:$res (Clift_ImmediateOp $value), $kind),
+        (createImmediateOp (getType $res),
+                           (NativeCodeCall<"truncateIntegerValue($0, $1)"> $value, $res)),
+        [(Clift_AnyPrimitiveIntegerType $res)]>;
+
+// (enum E)123 -> A
+// Constraint: A is an enumerator of E whose value is 123.
+def EnumImmediatePattern
+  : Pat<(Clift_CastOp:$res (Clift_ImmediateOp $value), $kind),
+        (createImmediateOp (getType $res), $value),
+        [(IsBitcastKind $kind),
+         (Constraint<CPred<[{ hasEnumeratorValue($0.getType(),
+                                                 $1.getValue().getZExtValue()) }]>> $res, $value)]>;
+
+// *&x -> x
+def IndirectionAddressofPattern
+  : Pat<(Clift_IndirectionOp (Clift_AddressofOp $arg)),
+        (replaceWithValue $arg)>;
+
+// &*x -> x
+def AddressofIndirectionPattern
+  : Pat<(Clift_AddressofOp (Clift_IndirectionOp $arg)),
+        (replaceWithValue $arg)>;
+
+// (T)x -> x
+// Constraint: the type of x is equivalent to T.
+def RedundantCastPattern
+  : Pat<(Clift_CastOp:$res $arg, $kind),
+        (replaceWithValue $arg),
+        [(TypesAreEquivalent $res, $arg)]>;
+
+// (T)(U)x -> (T)x
+// Constraint: the cast kinds are the same.
+// Constraint: the cast kind is collapsible (implemented in .cpp file).
+def CollapsibleCastPattern
+  : Pat<(Clift_CastOp:$res (Clift_CastOp $arg, $inner_kind), $outer_kind),
+        (Clift_CastOp $arg, $inner_kind),
+        [(ValuesAreEqual $inner_kind, $outer_kind),
+         (Constraint<CPred<"isCollapsibleCastKind($0.getValue())">> $inner_kind)]>;
+
+// *(T*)p = x ->         *p = (T)x
+//            -> *(T*)&( *p = (T)x )
```

Comment:
This rewrite is worse than the original form.
The problem is that you can't put a constraint on the fact that the assignment is discarded.
Just drop this rewrite from the PR and from the associated REP.